src/Http/Routing/src/CompositeEndpointDataSource.cs

```diff
@@ -87,137 +70,204 @@ public override IReadOnlyList<Endpoint> Endpoints
-                    {
-                        sb.Append(", Http Methods: ");
-                        sb.AppendJoin(", ", httpMethodMetadata.HttpMethods);
-                    }
-                    sb.Append(", Display Name: ");
-                    sb.Append(routeEndpoint.DisplayName);
-                    sb.AppendLine();
-                }
-                else
-                {
-                    sb.Append("Non-RouteEndpoint. DisplayName:");
-                    sb.AppendLine(endpoint.DisplayName);
-                }
+                // We have to hook to any OnChange callbacks before caching endpoints,
+                // otherwise we might miss changes that occurred to one of the _dataSources after caching.
+                CreateChangeTokenUnsynchronized(collectionChanged);
+            }
+
+            // Don't update endpoints if no one has read them yet.
+            if (_endpoints is not null)
+            {
+                // Refresh the endpoints from data source so that callbacks can get the latest endpoints.
+                CreateEndopintsUnsynchronized();
             }
-            return sb.ToString();
+        }
+
+        // Disposing registrations can block on user defined code on running on other threads that could try to acquire the _lock.
+        if (collectionChanged && oldChangeTokenRegistrations is not null)
+        {
+            foreach (var registration in oldChangeTokenRegistrations)
+            {
+                registration.Dispose();
+            }
+        }
+
+        // Raise consumer callbacks. Any new callback registration would happen on the new token created in earlier step.
+        // Avoid raising callbacks inside a lock.
+        oldTokenSource?.Cancel();
+    }
+
+    [MemberNotNull(nameof(_consumerChangeToken))]
+    private void CreateChangeTokenUnsynchronized(bool collectionChanged)
+    {
+        _cts = new CancellationTokenSource();
+        _consumerChangeToken = new CancellationChangeToken(_cts.Token);
 
-            static IEnumerable<string> FormatValues(IEnumerable<KeyValuePair<string, object?>> values)
+        if (collectionChanged)
+        {
+            _changeTokenRegistrations = new();
+            foreach (var dataSource in _dataSources)
             {
-                return values.Select(
-                    kvp =>
-                    {
-                        var value = "null";
-                        if (kvp.Value != null)
-                        {
-                            value = "\"" + kvp.Value.ToString() + "\"";
-                        }
-                        return kvp.Key + " = " + value;
-                    });
+                _changeTokenRegistrations.Add(ChangeToken.OnChange(
+                    dataSource.GetChangeToken,
+                    () => HandleChange(collectionChanged: false)));
             }
         }
     }
+
+    [MemberNotNull(nameof(_endpoints))]
+    private void CreateEndopintsUnsynchronized()
```

Comment:
nit: typo

```suggestion
    private void CreateEndpointsUnsynchronized()
```

Also don't merge this suggestion because you need to change the call site too.